### PR TITLE
Fix #11236 - CsvParser and ScheduleFile handling of edge cases to avoid crash

### LINKED ISSUE #11236: Schedule:File crash
### Issue overview

User Schedule:File object is not consistent with the csv file resulting in an arraybounds crash while reading the file.
```
Schedule:File,
    Rainfall Data from EPW file,  !- Name
    Fractional,              !- Schedule Type Limits Name
    SAU_MK_Mecca.410300_TMYx.csv,  !- File Name
    1,                       !- Column Number
    0,                       !- Rows to Skip at Top
    8760,                    !- Number of Hours of Data
    ,                        !- Column Separator
    ,                        !- Interpolate to Timestep
    ;                        !- Minutes per Item
```

csv input:

Image

Setting Schedule:File to this avoids the crash (but then the idf encounters other fatal errors).
```
    31,                       !- Column Number
    19,                       !- Rows to Skip at Top
```

### Operating System (Multiple choices)

Any

### Operating System Version

any

### Version of EnergyPlus

25.2.0-08aab32300

### Unmethours link or helpdesk ticket number

Helpdesk ticket 16954

### Defect file

EnergyplusDevSupport\DefectFiles

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

- Fixes #11236

### Description of the purpose of this PR

- Improve CsvParser:
    - skip extra columns (but register a warning)
    - Emplace a null when two consecutive delimiters found (but register a warning)
    - Neither of these issues are fatal yet, if you aren't actually requesting a column that is passed the parsed one or trying to get a number from the null column, you're just fine for now
- Improve ScheduleManager:
     - print the new warnings
     - Gracefully Fatal:
         - if trying to dereference a column with a null
         - if trying to access a column that is Past the number of columns we parsed (instead of throwing a nlohmann json crypic message) 
     - Schedule:File:Shading: fatal out in the edge case where the number of headers is > to the number of actual parsed columns. 
- Add lots of tests for ScheduleManager, and a new test file CsvParser.unit.cc to test out lower level functionality 

### Pull Request Author

 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer

 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/InputProcessing/CsvParser.cc [hunk 1]
```diff
@@ -64,6 +64,16 @@ bool CsvParser::hasErrors()
...
 
+std::vector<std::pair<std::string, bool>> const &CsvParser::warnings()
+{
+    return warnings_;
+}
+
+bool CsvParser::hasWarnings()
+{
+    return !warnings_.empty();
+}
```
Comment (jmarrec) [inline src/EnergyPlus/InputProcessing/CsvParser.cc:75]: Putting back warnings in CsvParser

File: src/EnergyPlus/InputProcessing/CsvParser.cc [hunk 2]
```diff
@@ -212,33 +222,71 @@ void CsvParser::parse_line(std::string_view csv, size_t &index, json &columns)
...
             ++column_num;
         } else {
-            columns.at(column_num).push_back(parse_value(csv, index));
+            if (column_num < num_columns) {
+                columns.at(column_num).push_back(parse_value(csv, index));
+            } else {
+                // Just parse and ignore the value
+                parse_value(csv, index);
+                has_extra_columns = true;
+            }
```
Comment (jmarrec) [inline src/EnergyPlus/InputProcessing/CsvParser.cc:308]: Avoid crashing here if you end up finding more values on the row than the number of columns we've determined by parsing the first data row (after header if present)

We setup the has_extra_columns to true here, so we can register a warning

File: src/EnergyPlus/InputProcessing/CsvParser.cc [hunk 3]
```diff
@@ -212,33 +222,71 @@ void CsvParser::parse_line(std::string_view csv, size_t &index, json &columns)
...
-                std::string line;
-                if (found_index != std::string::npos) {
-                    line = csv.substr(this_beginning_of_line_index, found_index - this_beginning_of_line_index);
-                }
                 errors_.emplace_back(
                     fmt::format(
                         "CsvParser - Line {} - Expected {} columns, got {}. Error in following line.", this_cur_line_num, num_columns, parsed_values),
                     false);
-                errors_.emplace_back(line, true);
+                errors_.emplace_back(getCurrentLine(), true);
```
Comment (jmarrec) [inline src/EnergyPlus/InputProcessing/CsvParser.cc]: When we reach the end of the line, we issue a warning if has_extra_columns, **otherwise** an error if the resulting number of parsed values is not the expected one.

File: src/EnergyPlus/InputProcessing/CsvParser.cc [hunk 4]
```diff
@@ -212,33 +222,71 @@ void CsvParser::parse_line(std::string_view csv, size_t &index, json &columns)
...
+                    warnings_.emplace_back(fmt::format("CsvParser - Line {} Column {} - Blank value found, setting to null. Error in following line.",
+                                                       this_cur_line_num,
+                                                       next_col + 1),
+                                           false);
+                    warnings_.emplace_back(getCurrentLine(), true);
+                } else {
+                    has_extra_columns = true;
+                }
+                ++parsed_values;
+            }
```
Comment (jmarrec) [inline src/EnergyPlus/InputProcessing/CsvParser.cc:299]: In the Delimiter case, we scan ahead to check if another delimiter is coming up, in which case we emplace a null.

This is a warning, because unless you actually try to use that column, it's fine.

File: src/EnergyPlus/ScheduleManager.cc [hunk 5]
```diff
@@ -615,6 +615,13 @@ namespace Sched {
                         ShowContinueError(state, fmt::format("Error Occurred in {}", state.files.TempFullFilePath.filePath));
                         ShowFatalError(state, "Program terminates due to previous condition.");
                     }
+                    for (const auto &[warning, isContinued] : csvParser.warnings()) {
+                        if (isContinued) {
+                            ShowContinueError(state, warning);
+                        } else {
+                            ShowWarningError(state, warning);
+                        }
+                    }
```
Comment (jmarrec) [inline src/EnergyPlus/ScheduleManager.cc:624]: Schedule:File:Shading, print the new warnings if any.

File: src/EnergyPlus/ScheduleManager.cc [hunk 6]
```diff
@@ -1643,12 +1650,19 @@ namespace Sched {
...
                             ShowContinueError(state, fmt::format("Error Occurred in {}", state.files.TempFullFilePath.filePath));
                             ShowFatalError(state, "Program terminates due to previous condition.");
                         }
+                        for (const auto &[warning, isContinued] : csvParser.warnings()) {
+                            if (isContinued) {
+                                ShowContinueError(state, warning);
+                            } else {
+                                ShowWarningCustom(state, eoh, warning);
+                            }
+                        }
```
Comment (jmarrec) [inline src/EnergyPlus/ScheduleManager.cc:1665]: Schedule:File, same; print warnings if any

File: src/EnergyPlus/ScheduleManager.cc [hunk 7]
```diff
@@ -1663,9 +1677,25 @@ namespace Sched {
                     }
                 }
 
+                // curcolCount is 1-indexed here
+                if (static_cast<size_t>(curcolCount) > result->second["values"].size()) {
+                    ShowSevereCustom(
+                        state, eoh, format("Requested column number {}, but found only {} columns.", curcolCount, result->second["values"].size()));
+                    ShowContinueError(state, fmt::format("Error Occurred in {}", state.files.TempFullFilePath.filePath));
+                    ShowFatalError(state, "Program terminates due to previous condition.");
+                }
```
Comment (jmarrec) [inline src/EnergyPlus/ScheduleManager.cc:1686]: Schedule:File: handle the case where I might request column 10000 from a CSV file that actually has 2 columns.

Otherwise you get a cryptic json error

File: src/EnergyPlus/ScheduleManager.cc [hunk 8]
```diff
@@ -1663,9 +1677,25 @@ namespace Sched {
...
+
+                std::vector<Real64> column_values;
+                try {
+                    column_values = column_json.get<std::vector<Real64>>();
+                } catch (nlohmann::json::type_error &e) {
+                    ShowSevereCustom(state, eoh, format("Column number {} has non-numeric data.", curcolCount));
+                    ShowContinueError(state, e.what());
+                    ShowContinueError(state, fmt::format("Error Occurred in {}", state.files.TempFullFilePath.filePath));
+                    ShowFatalError(state, "Program terminates due to previous condition.");
+                }
```
Comment (jmarrec) [inline src/EnergyPlus/ScheduleManager.cc:1698]: Gracefully handle using a column that had a null (or a string for that matter) instead of throwing a crytpic json error

File: src/EnergyPlus/ScheduleManager.cc [hunk 9]
```diff
@@ -1761,6 +1791,9 @@ namespace Sched {
             auto const headers = schedule_file_shading_result->second["header"].get<std::vector<std::string>>();  // (AUTO_OK_OBJ)
             auto const headers_set = schedule_file_shading_result->second["header"].get<std::set<std::string>>(); // (AUTO_OK_OBJ)
 
+            std::string const shadingFileName = schedule_file_shading_result->first.filename().string();
+            ErrorObjectHeader eoh{routineName, "Schedule:File:Shading", shadingFileName};
```
Comment (jmarrec) [inline src/EnergyPlus/ScheduleManager.cc:1795]: The Schedule:File:Shading parsing is done much earlier, here it's the bit of code that actually try to process the data.

File: src/EnergyPlus/ScheduleManager.cc [hunk 10]
```diff
@@ -1770,7 +1803,26 @@ namespace Sched {
...
+
+                if (column >= values_json.size()) {
+                    ShowSevereCustom(
+                        state,
+                        eoh,
+                        fmt::format(
+                            "For header '{}', Requested column number {}, but found only {} columns.", header, column + 1, values_json.size()));
+                    ShowContinueError(state, fmt::format("Error Occurred in {}", schedule_file_shading_result->first));
+                    ShowFatalError(state, "Program terminates due to previous condition.");
+                }
```
Comment (jmarrec) [inline src/EnergyPlus/ScheduleManager.cc:1815]: Schedule:File:Shading: handle the case where I have more headers than data columns

File: src/EnergyPlus/ScheduleManager.cc [hunk 11]
```diff
@@ -1770,7 +1803,26 @@ namespace Sched {
...
+
+                std::vector<Real64> column_values;
+                try {
+                    column_values = values_json.at(column).get<std::vector<Real64>>();
+                } catch (nlohmann::json::type_error &e) {
+                    ShowSevereCustom(state, eoh, fmt::format("Column number {} has non-numeric data.", column + 1));
+                    ShowContinueError(state, e.what());
+                    ShowContinueError(state, fmt::format("Error Occurred in {}", schedule_file_shading_result->first));
+                    ShowFatalError(state, "Program terminates due to previous condition.");
+                }
```
Comment (jmarrec) [inline src/EnergyPlus/ScheduleManager.cc:1825]: Shading: error gracefully if there is non numeric data in that column

File: tst/EnergyPlus/unit/CsvParser.unit.cc [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+        auto const &col = values[2];
+        EXPECT_EQ(2, col.size());
+        EXPECT_EQ(0.01, col[0]);
+        EXPECT_EQ(0.02, col[1]);
+    }
+
+    EXPECT_NO_THROW(values.at(0).get<std::vector<Real64>>());
+    EXPECT_NO_THROW(values.at(1).get<std::vector<Real64>>());
+    EXPECT_NO_THROW(values.at(2).get<std::vector<Real64>>());
+}
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/CsvParser.unit.cc:275]: Example test for CsvParser itself. It shouldn't crash if you have extra columns, just warn

File: tst/EnergyPlus/unit/ScheduleManager.unit.cc [hunk 13]
```diff
@@ -1492,20 +1500,267 @@ TEST_F(EnergyPlusFixture, ScheduleFile_Blanks)
...
+TEST_F(EnergyPlusFixture, ScheduleFile_Blanks_OnlyWarnIfNotUsingThatColumn)
+{
+    // On the third line (second data record after header), there is a blank in the second column
+    // Hour,Value1,Value2
+    // 0,0.01,0.01
+    // 1,,0.33
+    // 2,0.37,0.37
+    fs::path scheduleFile = FileSystem::makeNativePath(configured_source_directory() / "tst/EnergyPlus/unit/Resources/schedule_file_with_blank.csv");
+
+    // Here I am requested a column that is properly filled, and it should work fine
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ScheduleManager.unit.cc:1527]: New test. it works if I have a blank value but I don't try to use that column specifically.

File: tst/EnergyPlus/unit/ScheduleManager.unit.cc [hunk 14]
```diff
@@ -1492,20 +1500,267 @@ TEST_F(EnergyPlusFixture, ScheduleFile_Blanks)
...
+{
+    // On the third line (second data record after header), there is a blank in the second column, no extra delimiter.
+    // Hour,Value1
+    // 0,0.01
+    // 1,
+    // 2,0.37
+    fs::path scheduleFile =
+        FileSystem::makeNativePath(configured_source_directory() / "tst/EnergyPlus/unit/Resources/schedule_file_with_missing_value.csv");
+
+    // In this case, the csvParser registers an error and that one is thrown
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ScheduleManager.unit.cc:1576]: I have a completely missing value here, so I throw

File: tst/EnergyPlus/unit/ScheduleManager.unit.cc [hunk 15]
```diff
@@ -1492,20 +1500,267 @@ TEST_F(EnergyPlusFixture, ScheduleFile_Blanks)
...
+    // On the third line (second data record after header), there is an extra column
+
+    // Hour,Value1,
+    // 0,0.01,
+    // 1,0.04,0.33
+    // 2,0.37,
+    fs::path scheduleFile =
+        FileSystem::makeNativePath(configured_source_directory() / "tst/EnergyPlus/unit/Resources/schedule_file_with_extra_column.csv");
+
+    // I am requesting column 2, so it should warn but work
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ScheduleManager.unit.cc:1631]: I have an extra column, I warn and move on since it's not being used.

File: tst/EnergyPlus/unit/ScheduleManager.unit.cc [hunk 16]
```diff
@@ -1492,20 +1500,267 @@ TEST_F(EnergyPlusFixture, ScheduleFile_Blanks)
...
+    s_glob->MinutesInTimeStep = 15; // must initialize this to get schedules initialized
+    s_glob->TimeStepZone = 0.25;
+    s_glob->TimeStepZoneSec = s_glob->TimeStepZone * Constant::rSecsInHour;
+    state->dataEnvrn->CurrentYearIsLeapYear = false;
+
+    EXPECT_THROW(state->init_state(*state), EnergyPlus::FatalError); // read schedules
+
+    const std::string expected_error = delimited_string({
+        "   ** Severe  ** ProcessScheduleInput: Schedule:File = TEST1",
+        "   **   ~~~   ** Requested column number 100, but found only 2 columns.",
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ScheduleManager.unit.cc:1715]: Requesting a column that doesn't exist: fatal

File: tst/EnergyPlus/unit/ScheduleManager.unit.cc [hunk 17]
```diff
@@ -1492,20 +1500,267 @@ TEST_F(EnergyPlusFixture, ScheduleFile_Blanks)
...
+        "   ** Severe  ** ProcessScheduleInput: Schedule:File:Shading = shading_data_2220_broken.csv",
+        "   **   ~~~   ** For header 'WEST SIDE TREE', Requested column number 3, but found only 2 columns.",
+        "   **   ~~~   ** Error Occurred in " + scheduleFile.string(),
+        "   **  Fatal  ** Program terminates due to previous condition.",
+        "   ...Summary of Errors that led to program termination:",
+        "   ..... Reference severe error count=1",
+        "   ..... Last severe error=ProcessScheduleInput: Schedule:File:Shading = shading_data_2220_broken.csv",
+    });
+    compare_err_stream(expected_error);
+}
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ScheduleManager.unit.cc:1763]: Test edge case for Schedule:File:Shading where you have more headers than data columns

**Comment by jmarrec** [comment]
Hum, this is conflicted. Going to rebase, hopefully I don't botch it

**Review by mitchute — APPROVED** [review]
